CIP-0037/CIP-0037.md

```diff
@@ -0,0 +1,338 @@
+Let k be the number of submitted signatures (as above, k is a security parameter determining the number of 
+required signatures). However, in mithril k << N where N is the total number of eligible signers. Mithril 
+verification goes as foollows: 
+* Check the k is large enough
+* k G_1 additions
+* k G_2 additions
+* k (log_2(N) hash computations + equality checks)
+* 2 pairing evaluations + equality check
+
+There are some additional t checks of lottery winning (similar to VRF). These are non crytpgraphic operations 
+(rather inequality checks of indices, and checks that some hash computation is below a threshold).
+
+## Coconut certs
+We use this as an example for anonymous credentials. There is no decision on what type of construction we will 
+eventually use. Coconut credentials (and other types of credentials we are looking at) are built in a way that 
+it is efficient to prove statements about attributes. To this end, it is required to build, and be able to verify, 
+relations over discrete logarithm values. However, for sake of simplicity in this computation, we consider the 
+simplest form of anonymous credentials, which contains no attributes. We not that proving statements regarding 
+attributes does not require further pairing evaluations. 
+* Proof verification to a relation involving elements in G_1 and G_2. This verification does not require the 
+* pairing evaluation, but does require G_1 and G_2 additions and multiplications. 
+* 2 pairing evaluations + equality check
+
+The complexity of the statement one can prove with anonymous credentials will be bounded by plutus computational 
+limits. However, it is hard to make individual functions for each possible statement accepted. 
+
+## Plonk verification
+Plonk verification is quite complex, and might differ depending on the number of custom gates, or specific wire 
+connections. In this section we expose the verifier complexity as described 
+[here](https://hackmd.io/Vr0DUYPCR0iWADpO8kpkAA?view#Verifier-Algorithm). We count every challenge computation 
+as a single hash evaluation. We omit the `scalar X scalar` multiplications and additions. As we can see above, 
+these operations are insignificant with respect to group operations. We assume that point validation is performed 
+by the external C library, and therefore do not count that in this estimate. The computation is the following: 
+
+* 6 hash computations
+* 1 vanishing polynomial (1 scalar exponantiation)
+* Quotient polynomial evaluation (1 scalar inverse)
+* First part of batch poly commitment (7 scalar mults and 6 additions in G_1)
+* Fully batched poly commitment (7 scalar mults and 8 additions over G_1)
+* Group encoded batch verification (1 scalar mult over G_1)
+* Batch-validate all evaluations (3 scalar mults and 4 additions over G_1, plus two pairing evaluations + equality check)
+
+
+Implementations may differ, and adding custom gates or additional wires will affect these estimates. Pairing 
```

Comment:
This is a good estimate as a reference, but we have to expect higher costs because this is the most basic version of Plonk and real life implementations will likely be more complex. 
For example adding lookups gates introduces tables (with its corresponding commitments), an extra permutation argument (increases the number of scalar multiplications and additions) and the need for new challenges (more hashes).

Reply:
Blinder costs should also be considered as well as the custom gates that any implementation uses.

From my perspective, plonk is "completely agnostic to this PR". In the sense that any project will be free to implement it's own Plonk Verifier contract/lib with their own custom gate design picks for their use-cases.
What I think would be really valuable is to first make sure that this is indeed doable for `vanilla-plonk` or `turbo-plonk` (excluding custom gates) as we still don't have any idea of how much this execution will cost or even if it is possible to execute all that stuff inside a contract.

Reply:
Thanks. Indeed, as you say this CIP is agnostic to what the pairings are used for. This list served as an estimate of some of the use cases that could leverage pairings. On your second point, I complete agree as well. Once we get the pairings running on a test version of plutus (and we have the functions costed), we'll see how it performs for verifying a vanilla plonk, and how much of the execution budget we spend.